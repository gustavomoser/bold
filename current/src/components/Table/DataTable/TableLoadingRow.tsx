import React, { CSSProperties } from 'react'

import { Theme, useStyles } from '../../../styles'
import { Spinner } from '../../Spinner'
import { TableCell, TableRow } from '../Table'

export interface TableLoadingRowProps {
  colSpan: number
  message?: string
}

export function TableLoadingRow(props: TableLoadingRowProps) {
  const { colSpan, message } = props
  const { classes } = useStyles(createStyles)

  return (
    <TableRow>
      <TableCell colSpan={colSpan} style={classes.cell}>
        <div className={classes.container}>
          <Spinner />
          {message}
        </div>
      </TableCell>
    </TableRow>
  )
}

TableLoadingRow.defaultProps = {
  message: 'Carregando resultados',
} as Partial<TableLoadingRowProps>

const createStyles = (theme: Theme) => ({
  cell: {
    background: theme.pallete.surface.background,
    color: theme.pallete.primary.main,
    fontWeight: 'bold',
  } as CSSProperties,
  container: {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    '& > *:first-of-type': {
      marginRight: '0.5rem',
    },
  } as CSSProperties,
})
