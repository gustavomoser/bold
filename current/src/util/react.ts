import { MutableRefObject, Ref } from 'react'

/**
 * Compose multiple react refs.
 *
 * @returns A new ref that passes down the Element to all ref arguments.
 */
export function composeRefs<T>(...refs: Array<Ref<T>>) {
  return (value: T) => refs.forEach(ref => setRef(ref, value))
}

/**
 * Changes the value of a ref.
 * The ref can be a function ref or a object ref.
 *
 * @param ref The ref to have its value changed.
 * @param value The new value.
 */
export function setRef<T>(ref: Ref<T>, value: any) {
  if (typeof ref === 'function') {
    ref(value)
  } else if (ref) {
    const mutableRef = ref as MutableRefObject<T>
    mutableRef.current = value
  }
}

/**
 * Compose multiple event handlers.
 *
 * @returns A new function that triggers all handlers passed as parameters, in order.
 */
export function composeHandlers(...handlers: Array<(...params: any) => void>) {
  return (...args: any[]) => handlers.forEach(handler => handler && handler(...args))
}

/**
 * Invokes the native `value` property setter of an element.
 * From https://github.com/facebook/react/issues/10135#issuecomment-401496776
 *
 * @param element The element to invoke the setter on.
 * @param value The value to be set.
 */
export function setNativeValue(element: HTMLElement, value: any) {
  const prototype = Object.getPrototypeOf(element)
  const { set: prototypeValueSetter = null } = Object.getOwnPropertyDescriptor(prototype, 'value') || {}
  const { set: valueSetter = null } = Object.getOwnPropertyDescriptor(element, 'value') || {}

  if (prototypeValueSetter && valueSetter !== prototypeValueSetter) {
    prototypeValueSetter.call(element, value)
  } else if (valueSetter) {
    valueSetter.call(element, value)
  } else {
    throw new Error('The given element does not have a value setter')
  }
}
