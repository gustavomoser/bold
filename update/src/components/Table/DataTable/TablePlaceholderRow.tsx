import React, { CSSProperties } from 'react'

import { Theme, useStyles } from '../../../styles'
import { TableCell, TableRow } from '../Table'

export interface TablePlaceholderRowProps {
  colSpan: number
  message?: string
}

export function TablePlaceholderRow(props: TablePlaceholderRowProps) {
  const { colSpan, message } = props
  const { classes } = useStyles(createStyles)

  return (
    <TableRow>
      <TableCell colSpan={colSpan} style={classes.cell}>
        {message}
      </TableCell>
    </TableRow>
  )
}

TablePlaceholderRow.defaultProps = {
  message: 'Nenhum registro encontrado',
} as Partial<TablePlaceholderRowProps>

export const createStyles = (theme: Theme) => ({
  cell: {
    color: theme.pallete.text.secondary,
    fontStyle: 'italic',
    textAlign: 'center',
  } as CSSProperties,
})
