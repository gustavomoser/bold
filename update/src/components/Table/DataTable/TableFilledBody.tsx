import React from 'react'

import { useCss } from '../../../styles'
import { TableBody, TableCell, TableRow } from '../Table'

import { DataTableProps, defaultColumnStyles } from './DataTable'
import { TableLoadingRow } from './TableLoadingRow'
import { TablePlaceholderRow } from './TablePlaceholderRow'

export interface TableFilledBodyProps extends Pick<DataTableProps, 'columns' | 'rows' | 'loading' | 'onRowClick'> {}

export function TableFilledBody(props: TableFilledBodyProps) {
  const { columns, rows, loading, onRowClick } = props
  const { css } = useCss()

  const handleClick = row => e => onRowClick(row)

  const isEmpty = () => !rows || rows.length === 0

  return (
    <TableBody>
      {loading && <TableLoadingRow colSpan={columns.length} />}

      {!loading && isEmpty() && <TablePlaceholderRow colSpan={columns.length} />}

      {rows.map((row, idx) => (
        <TableRow key={idx} onClick={onRowClick && handleClick(row)}>
          {columns.map((col, colIdx) => (
            <TableCell key={colIdx} style={css(defaultColumnStyles(col), col.style)}>
              {col.render(row)}
            </TableCell>
          ))}
        </TableRow>
      ))}
    </TableBody>
  )
}
