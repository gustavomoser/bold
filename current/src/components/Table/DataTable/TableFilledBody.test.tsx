import { render } from '@testing-library/react'
import React from 'react'

import { TableFilledBody, TableFilledBodyProps } from './TableFilledBody'
import { TableLoadingRow } from './TableLoadingRow'
import { TablePlaceholderRow } from './TablePlaceholderRow'

interface Row {
  id: number
  name: string
}

const rows: Row[] = [
  { id: 1, name: 'MARIA MACHADO DE JESUS' },
  { id: 2, name: 'JOSÃ‰ DA SILVA MOREIRA' },
  { id: 3, name: 'ALICE BARBOSA' },
]

const createTable = (props?: Partial<TableFilledBodyProps>) => (
  <table>
    <TableFilledBody
      columns={[
        { name: 'id', header: 'ID', sortable: true, render: (row: Row) => row.id },
        { name: 'name', header: 'Name', sortable: true, render: (row: Row) => row.name },
      ]}
      rows={rows}
      loading={false}
      {...props}
    />
  </table>
)

it('should render correctly', () => {
  const { container } = render(createTable())
  expect(container).toMatchSnapshot()
})

it('should render the TableLoadingRow if loading is true', () => {
  const queryText = TableLoadingRow.defaultProps.message
  expect(render(createTable({ loading: false })).queryAllByText(queryText).length).toEqual(0)
  expect(render(createTable({ loading: true })).queryAllByText(queryText).length).toEqual(1)
})

it('should render TablePlaceholderRow if loading is false and there are no rows', () => {
  const queryText = TablePlaceholderRow.defaultProps.message
  expect(render(createTable()).queryAllByText(queryText).length).toEqual(0)
  expect(render(createTable({ rows: [], loading: true })).queryAllByText(queryText).length).toEqual(0)
  expect(render(createTable({ rows: [], loading: false })).queryAllByText(queryText).length).toEqual(1)
})
