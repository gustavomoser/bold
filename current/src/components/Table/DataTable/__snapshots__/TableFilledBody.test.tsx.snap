// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
.emotion-0 {
  text-align: left;
  padding: 0 1rem;
  height: 2.5rem;
  border-bottom: 1px solid #D3D4DD;
}

<div>
  <table>
    <tbody
      class="emotion-2"
    >
      <tr
        class="emotion-2"
      >
        <td
          class="emotion-0"
        >
          1
        </td>
        <td
          class="emotion-0"
        >
          MARIA MACHADO DE JESUS
        </td>
      </tr>
      <tr
        class="emotion-2"
      >
        <td
          class="emotion-0"
        >
          2
        </td>
        <td
          class="emotion-0"
        >
          JOSÃ‰ DA SILVA MOREIRA
        </td>
      </tr>
      <tr
        class="emotion-2"
      >
        <td
          class="emotion-0"
        >
          3
        </td>
        <td
          class="emotion-0"
        >
          ALICE BARBOSA
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
